contracts/v2/MarketWithAuction.sol

```diff
@@ -0,0 +1,902 @@
+    function acceptOffer(
+        uint256 _listingId, 
+        address offeror
+    ) 
+        external
+        override
+        nonReentrant
+        onlyListingCreator(_listingId)
+    {
+        Offer memory targetOffer = offers[_listingId][offeror];
+        Listing memory targetListing = listings[_listingId];
+
+        require(
+            targetListing.listingType == ListingType.Direct
+                && validateOwnershipAndApproval(
+                    targetListing.tokenOwner, 
+                    targetListing.assetContract, 
+                    targetListing.tokenId, 
+                    targetOffer.quantityWanted, 
+                    targetListing.tokenType,
+                    targetListing.listingType
+                ), 
+            "Market: must own and approve to transfer tokens."
+        );
+
+        uint256 quantityWanted = targetOffer.quantityWanted;
+        uint256 offerAmount = targetOffer.pricePerToken * targetOffer.quantityWanted;
+
+        targetOffer.quantityWanted = 0;
+        targetOffer.pricePerToken = 0;
+        offers[_listingId][offeror] = targetOffer;
+
+        targetListing.quantity -= quantityWanted;
+        listings[_listingId] = targetListing;
+
+        payout(
+            offeror, 
+            targetListing.tokenOwner,
+            targetOffer.currency,
+            offerAmount, 
+            targetListing
+        );
+        transferListingTokens(targetListing.tokenOwner, offeror, quantityWanted, targetListing);
+
+        emit NewDirectSale(
+            targetListing.assetContract, 
+            targetListing.tokenOwner, 
+            _listingId, 
+            offeror, 
+            quantityWanted,
+            targetListing
+        );
+    }
+
+    /// @dev Lets an auction's creator close the auction.
+    function closeAuction(
+        uint256 _listingId
+    )
+        external
+        override
+        nonReentrant
+    {
+        Listing memory targetListing = listings[_listingId];
+        Offer memory targetBid = winningBid[_listingId];
+        
+        address closer = _msgSender();
+
+        if(targetListing.startTime > block.timestamp) {
+
+            require(
+                listings[_listingId].tokenOwner == closer,
+                "Market: caller is not the listing creator."
+            );
+
+            // Auction is considered canceled if end time has passed.
+            uint256 quantityToSend = targetListing.quantity;
+            targetListing.quantity = 0;
+            targetListing.endTime = block.timestamp;
+            listings[_listingId] = targetListing;
+
+            transferListingTokens(address(this), targetListing.tokenOwner, quantityToSend, targetListing);
+
+            emit AuctionCanceled(_listingId, targetListing.tokenOwner, targetListing);
+
+            return;
+        }
+
+        require(
+            closer == targetListing.tokenOwner || closer == targetBid.offeror,
```

Comment:
does it really matter? since the output goes to either tokenOwner and offeror already.
if you're feeling generous, you could be closing auctions for fun :)